simple/app.go

```diff
@@ -0,0 +1,334 @@
+		internalKinds:      make(map[string]resource.Kind),
+		converters:         make(map[string]Converter),
+		cfg:                config,
+	}
+	for _, kind := range config.ManagedKinds {
+		err := a.ManageKind(kind)
+		if err != nil {
+			return nil, err
+		}
+	}
+	for gk, converter := range config.Converters {
+		a.RegisterKindConverter(gk, converter)
+	}
+	a.operator.AddController(a.informerController)
+	return a, nil
+}
+
+// ValidateManifest can be called with app.ManifestData to validate that the current configuration and managed kinds
+// fully cover the kinds and capabilities in the provided app.ManifestData. If the provided app.ManifestData
+// contains a kind or a capability for a kind/version that is not covered by the app's currently managed kinds,
+// an error will be returned. If the app's current managed kinds cover more than the provided app.ManifestData
+// indicates, no error will be returned.
+// This method can be used after initializing an app to verify it matches the loaded app.ManifestData from the app runner.
+func (a *App) ValidateManifest(manifest app.ManifestData) error {
+	for _, k := range manifest.Kinds {
+		if _, ok := a.converters[schema.GroupKind{Group: manifest.Group, Kind: k.Kind}.String()]; !ok && k.Conversion {
+			return fmt.Errorf("kind %s has conversion enabled but no converter is registered", k.Kind)
+		}
+		for _, v := range k.Versions {
+			kind, ok := a.kinds[gvk(manifest.Group, v.Name, k.Kind)]
+			if !ok {
+				return fmt.Errorf("kind %s/%s exists in manifest but is not managed by the app", k.Kind, v.Name)
+			}
+			if v.Admission.SupportsAnyValidation() && kind.Validator == nil {
+				return fmt.Errorf("kind %s/%s supports validation but has no validator", k.Kind, v.Name)
+			}
+			if v.Admission.SupportsAnyMutation() && kind.Mutator == nil {
+				return fmt.Errorf("kind %s/%s supports mutation but has no mutator", k.Kind, v.Name)
+			}
+		}
+	}
+	return nil
+}
+
+// ManagedKinds returns a slice of all Kinds managed by this App
+func (a *App) ManagedKinds() []resource.Kind {
+	kinds := make([]resource.Kind, 0)
+	for _, k := range a.kinds {
+		kinds = append(kinds, k.Kind)
+	}
+	return kinds
+}
+
+// Runner returns a resource.Runnable() that runs the underlying operator.InformerController and all custom runners
+// added via AddRunnable. The returned resource.Runnable also implements metrics.Provider, allowing the caller
+// to gather prometheus.Collector objects used by all underlying runners.
+func (a *App) Runner() app.Runnable {
+	return a.operator
+}
+
+// AddRunnable adds an arbitrary resource.Runnable runner to the App, which will be encapsulated as part of Runner().Run().
+// If the provided runner also implements metrics.Provider, PrometheusCollectors() will be called when called on Runner().
+func (a *App) AddRunnable(runner app.Runnable) {
+	a.operator.AddController(runner)
+}
+
+// ManageKind introduces a new kind to manage.
+func (a *App) ManageKind(kind AppManagedKind) error {
```

Comment:
What is the use-case for this method? Why would someone use this instead of passing all kinds to the `NewApp` instead?

Reply:
Convenience, mainly. There have been some comments on the large config object getting unwieldy, though the contents of the config vs this method are the same thing, so it's mostly about adding the option.

Reply:
Let's limit the options then. I see footguns potential here - due to synchronization / concurrency.